File: src/forensics/epistemic_narrowing_monitor.ts
export interface EpistemicNarrowingResult {
  token_diversity: number;
  repetition_ratio: number;
  narrowing_detected: boolean;
  summary: string;
}

export function monitorEpistemicNarrowing(text: string): EpistemicNarrowingResult {
  const tokens = text.toLowerCase().split(/\s+/).filter(Boolean);
  const unique = new Set(tokens);
  const diversity = tokens.length ? unique.size / tokens.length : 0;
  const repetitions = tokens.length - unique.size;
  const repetitionRatio = tokens.length ? repetitions / tokens.length : 0;
  const narrowingDetected = diversity < 0.4 || repetitionRatio > 0.35;
  return {
    token_diversity: Number(diversity.toFixed(3)),
    repetition_ratio: Number(repetitionRatio.toFixed(3)),
    narrowing_detected: narrowingDetected,
    summary: narrowingDetected
      ? "Epistemic narrowing detected: low lexical diversity and elevated repetition."
      : "No significant narrowing detected.",
  };
}

File: src/forensics/epistemic_narrowing_monitor.py
"""Epistemic narrowing monitor with token diversity heuristics."""

def monitor(text: str) -> dict:
    tokens = [t for t in text.lower().split() if t]
    if not tokens:
        return {"token_diversity": 0.0, "repetition_ratio": 0.0, "narrowing_detected": False}
    unique = set(tokens)
    diversity = len(unique) / len(tokens)
    repetition_ratio = (len(tokens) - len(unique)) / len(tokens)
    narrowing = diversity < 0.4 or repetition_ratio > 0.35
    return {
        "token_diversity": round(diversity, 3),
        "repetition_ratio": round(repetition_ratio, 3),
        "narrowing_detected": narrowing,
    }
